### PR TITLE
User consent screen

### PR DESCRIPTION
**Pull request by cocoStein**
# Summary
As milestone 2 will focus on the User experience with real users, it's important that test users know which data is used and what we do/plan doing with it. There is also the legal point of view - while it may not strictly apply here, it's better to have proper consent than nothing at all.

> [!NOTE]  
> To test this features it's better if you wipe your account from the firebase, so you have to redo the flow of creating an account, and this is were you'll see this screen. Right after choosing a username !

# What
When the user first logs in, they're presented with a screen listing the terms they must agree to before proceeding to the actual app.

- UI
  - New `BetaConsentScreen` - displays beta program terms and data collection agreement
  - Scrollable content with sections for data collection, photos, and location
  - Checkbox agreement with Agree/Decline buttons
  - Error handling with Snackbar
  - Loading states during consent save operation
  
- ViewModel
  - New `BetaConsentViewModel` - manages consent state and persistence
  - New `BetaConsentViewModelFactory` - provides proper dependency injection for SharedPreferences
  - Handles both Firebase (permanent storage) and SharedPreferences (local caching)
  - State flows for: `hasConsented`, `isLoading`, `isInitializing`, `consentSaved`, `error`
  - Functions for recording, clearing, and checking consent status
  
- Repository
  - New `Consent` data model with fields: `consentUuid`, `userId`, `timestamp`, `version`
  - New `ConsentRepository` interface with methods for CRUD operations
  - New `ConsentRepositoryFirestore` implementation with Firestore backend
  - New `ConsentRepositoryProvider` for singleton repository access
  - Added Firestore rules for consents collection (users can only access their own records)
  
- Navigation
  - Added `Screen.BetaConsent` to navigation routes
  - Modified registration flow to check consent status before navigating to Feed
  - Auto-navigation to Feed after successful consent
  - Decline redirects to Authentication screen
  
- Test
  - connected tests for `BetaConsentScreen`
  - unit tests for `BetaConsentViewModel`
  - Unit tests for `Consent` data model 

# Why
**User Trust & Transparency**: Real users in Milestone 2 deserve to know exactly what data we collect and how it's used. This builds trust and ensures informed participation.

**Legal Compliance**: While this is an academic project, implementing proper consent mechanisms demonstrates professional software development practices and prepares the app for potential real-world deployment.

**Data Ethics**: Explicitly stating data collection practices (photos, location, usage analytics) aligns with modern privacy standards and ethical development practices. As seen in CS-290 (Responsible software)

# Screenshots / Videos
[Screen_recording_20251104_220347.webm](https://github.com/user-attachments/assets/bf5b3617-036f-4b13-9af9-ee305431f66f)

# Checklist
- [x] Tests: added comprehensive unit tests for UI, ViewModel, and data model
  - `BetaConsentScreenTest.kt` - UI tests covering all screen elements and interactions
  - `BetaConsentViewModelTest.kt` - tests covering all ViewModel logic and edge cases
  - `ConsentTest.kt` -  tests for data model validation
- [x] Documentation: Kdoc comments added to all new classes and methods
- [x] Manual verification: app tested with new user flow
  - Sign in → Register → Consent screen appears → Agree → Navigate to Feed
  - Consent persists across app restarts (checked via SharedPreferences and Firebase)
  - Decline properly returns to Authentication screen
- [x] Integration: Updated end2end test to mock consent as already given
- [x] Firestore Rules: Added security rules for consents collection
- [x] Error handling: Proper error messages for network failures and authentication issues

> [!WARNING]  
> This PR exceeds the 1.5k lines limit, but I don't think it's doable to split it, as we would just have one PR with "useless" features that can't be properly tested.
> This PR is big mostly because of the tests and because of the `BetaConsentScreen`, which should be pretty easy to review even with that much lines. The logic is pretty straight forward.

### CODE REVIEW

File: app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      consentUuid = this.consentUuid,
+      userId = this.userId,
+      timestamp = this.timestamp,
+      version = this.version)
+}
+
+class ConsentRepositoryFirestore(private val db: FirebaseFirestore) : ConsentRepository {
+
+  /** Helper method to validate consent data */
+  private fun checkConsentData(consent: Consent): Consent? {
```
Comment (MarcK0909) [inline app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt]: I think you could also check the timestamp just in case no ?

File: app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+  override suspend fun addConsent(consent: Consent) {
+    val validConsent = checkConsentData(consent)
+    if (validConsent == null) {
+      Log.e(CONSENT_TAG, "Cannot add invalid consent")
+      return
+    }
+
+    try {
+      db.collection(CONSENT_COLLECTION_PATH)
```
Comment (MarcK0909) [inline app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt]: I noticed in ConsentRepositoryFirestore.kt at line 86-93, we're using .set() to save the consent record. This will overwrite any existing consent if addConsent() is called twice with the same UUID. If a user somehow triggers consent recording multiple times, we'll lose the original timestamp, which could be important for audit purposes.

Consider adding a check before saving:

Check if a consent already exists for this userId (using getConsentByUserId) and return early if found, OR
Use .set(consent.toDto(), SetOptions.merge()) to avoid full overwrites, OR
Check if the document exists first with .get().await() before calling .set()
I'd recommend checking by userId to ensure one consent per user, since that seems to be the business logic we want.

File: app/src/main/java/com/android/ootd/MainActivity.kt [hunk 3]
```diff
@@ -119,7 +130,38 @@ fun OOTDApp(
...
+                        // After registration, show beta consent if not already given
+                        if (betaConsentViewModel.getConsentStatus()) {
+                          navigationActions.navigateTo(Screen.Feed)
+                        } else {
+                          navigationActions.navigateTo(Screen.BetaConsent)
+                        }
+                      })
+                }
+                composable(Screen.BetaConsent.route) {
+                  val consentSaved by betaConsentViewModel.consentSaved.collectAsState()
```
[the remark this answers is not in the document]
Comment (cocoStein) [reply app/src/main/java/com/android/ootd/MainActivity.kt]: I think it's a good idea, so we don't collect the state when the app is in the background, Good catch !

File: firebase/firestore/firestore.rules [hunk 4]
```diff
@@ -83,6 +83,21 @@ service cloud.firestore {
              && (resource == null || resource.data.ownerId == request.auth.uid);
     }
 
+    // Consents: users can only read/write their own consent records
+    match /consents/{consentId} {
+      allow read: if isAuthed() &&
+                     (resource == null || resource.data.userId == request.auth.uid);
```
[the remark this answers is not in the document]
Comment (cocoStein) [reply firebase/firestore/firestore.rules]: Great question! Yes, this is intentional and necessary. The `resource == null` check handles the case where the document doesn't exist yet. If for example the database is wipped. 

In Firestore security rules:
- `resource` represents the document being accessed
- `resource == null` means the document doesn't exist in the database yet

Without this check, if a user tries to read a consent document that doesn't exist (e.g., checking if they've consented before creating their first consent), the read would be denied, which would break the flow.

The security is still maintained because:
1. We still check `isAuthed()` - user must be authenticated
2. When the resource DOES exist, we verify `resource.data.userId == request.auth.uid` - user can only read their own consent

Tell me what you think, but i think it works fine that way.

File: app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    if (userId.isBlank()) {
+      Log.e(CONSENT_TAG, "Cannot delete with blank userId")
+      return
+    }
+    try {
+      val snapshot =
+          db.collection(CONSENT_COLLECTION_PATH).whereEqualTo(USER_ID_FIELD, userId).get().await()
+
+      val batch = db.batch()
+      snapshot.documents.forEach { batch.delete(it.reference) }
```
[the remark this answers is not in the document]
Comment (MarcK0909) [reply app/src/main/java/com/android/ootd/model/consent/ConsentRepositoryFirestore.kt]: Thinking of it like this it does indeed seem correct so thank you for explaining

**Review by MarcK0909 — APPROVED** [review]
LGTM malaka
